### PR TITLE
Mumble: implement lock file mechanism for Windows.

### PR DESCRIPTION
**Pull request by mkrautz**
The RPC/DBus-based method for ensuring that only a single instance of
Mumble is running is racy.

If one of the processes has started it's RPC listener, it works fine.
But if both processes are started at almost the same time, it's very
likely that neither process has started listening for RPC. And the
exclusivity check fails.

In order to be completely sure that only a single Mumble instance is
running, we now take a lock file.

We open a lock file file in the data directory (for Windows, that's
%APPDATA%\Mumble). The file has sharing disabled, so as long as the
process that created the file is running, no other process can access
it.

If a second instance of Mumble is launched, and the file is inaccessible,
that instance terminates itself.

It is, of course, still possible to override the single instance
limitation by passing the '-m' flag to Mumble.

Fixes mumble-voip/mumble#1797

### CODE REVIEW

File: src/mumble/main.cpp [hunk 1]
```diff
@@ -292,6 +292,34 @@ int main(int argc, char **argv) {
 		}
 	}
 
+	// The code above this block is somewhat racy, in that it might not
+	// be possible to do RPC/DBus if two processes start at almost the
+	// same.
```
Comment (hacst) [inline src/mumble/main.cpp]: + time ?

**Comment by mkrautz** [comment]
@hacst PTAL at v2: https://github.com/mkrautz/mumble/commit/2f74996c9797a81034de71b88ac719fd6754c487

**Comment by hacst** [comment]
As we leak the handle and don't actively close it during shutdown I'm a bit worried we might make the self-restart functionality (```MUMBLE_EXIT_CODE_RESTART```) racy with this change. The window is very small so I'm not sure if it's a problem in practice. WDYT?

**Comment by mkrautz** [comment]
We'll need to close the handle before starting the new process.

**Comment by hacst** [comment]
Also might think about using mutexes (e.g. http://stackoverflow.com/questions/8799646/preventing-multiple-instances-of-my-application) instead. That way we don't need a physical file that another stupid process might accidentally hold a handle too (creating fun and confusing error possibilities for us).

**Comment by mkrautz** [comment]
Mutexes are tough because they are tied to the current desktop session -- where processes from different users may be running. As such, a mutex in the local namespace will cover all processes in the current session, even a Mumble instance running as another user.

That is why I chose the lock file approach instead.

See http://stackoverflow.com/a/16255816

**Comment by hacst** [comment]
Sounds reasonable.

**Comment by mkrautz** [comment]
@hacst PTAL at v3: https://github.com/mkrautz/mumble/commit/b71641d2dafbd2c642dbd886aec172c255a50687

**Comment by mkrautz** [comment]
@hacst PTAL at v4: https://github.com/mkrautz/mumble/commit/658f69a934df735ea0843f98545ad7335ce90027

**Comment by mkrautz** [comment]
Sorry, missing path() const.

**Comment by mkrautz** [comment]
PTAL at v5: https://github.com/mkrautz/mumble/commit/57b3bfc3c471ddd399e53b9f556c8f85a37742ab

**Comment by hacst** [comment]
V5 LGTM.

**Comment by mkrautz** [comment]
Landed v5 as d0ced44788eb49e636da944da061596943f367db.

**Pull request closed without merging**